daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/Validation.scala

```diff
@@ -4,395 +4,53 @@
-                )
-                if version1 == version2 &&
-                  templateId1 == templateId2 &&
-                  keyIsReplayedBy(key1, key2) &&
-                  result1 === result2 =>
-              loop(rest1, rest2, stack)
-            case (
-                  Node.NodeRollback(
-                    children1
-                  ),
-                  Node.NodeRollback(
-                    children2
-                  ),
-                ) =>
-              loop(
-                children1.iterator.to(LazyList),
-                children2.iterator.to(LazyList),
-                RollbackEntry(Rollback(nid1, rest1), Rollback(nid2, rest2)) :: stack,
-              )
-            case _ =>
-              Left(ReplayNodeMismatch(recorded, nid1, replayed, nid2))
-          }
-
-        case (LazyList(), LazyList()) =>
-          stack match {
-            case ExerciseEntry(Exercise(nid1, exe1, nids1), Exercise(nid2, exe2, nids2)) :: rest =>
-              if (resultIsReplayedBy(exe1.exerciseResult, exe2.exerciseResult))
-                loop(nids1, nids2, rest)
-              else
-                Left(ReplayNodeMismatch(recorded, nid1, replayed, nid2))
-            case RollbackEntry(Rollback(_, nids1), Rollback(_, nids2)) :: rest =>
-              loop(nids1, nids2, rest)
-            case Nil =>
-              Right(())
-          }
-
-        case (LazyList.cons(nid1, _), LazyList()) =>
-          Left(ReplayedNodeMissing(recorded, nid1, replayed))
-
-        case (LazyList(), LazyList.cons(nid2, _)) =>
-          Left(RecordedNodeMissing(recorded, replayed, nid2))
-
-      }
-
-    loop(recorded.roots.iterator.to(LazyList), replayed.roots.iterator.to(LazyList))
-
+    val normalizer = new Normalization[Nid, Cid]()
+    val recordedN = normalizer.normalizeTx(recorded)
+    val replayedN = normalizer.normalizeTx(replayed)
+    if (recordedN == replayedN) {
+      Right(())
+    } else {
+      Left(ReplayMismatch(recordedN, replayedN))
+    }
   }
-
 }
 
 object Validation {
-
+  //TODO: remove 'isReplayedBy' and instead have caller directly perform normalization + equality test
   def isReplayedBy[Nid, Cid](
       recorded: VersionedTransaction[Nid, Cid],
       replayed: VersionedTransaction[Nid, Cid],
-  )(implicit ECid: Equal[Cid]): Either[ReplayMismatch[Nid, Cid], Unit] =
+  ): Either[ReplayMismatch[Nid, Cid], Unit] =
     new Validation().isReplayedBy(recorded, replayed)
-
-  // package private for test.
-  private[lf] def valueIsReplayedBy[Cid](recorded: Value[Cid], replayed: Value[Cid])(implicit
-      ECid: Equal[Cid]
-  ): Boolean =
-    new Validation().valueIsReplayedBy(recorded, replayed)
-
 }
 
-sealed abstract class ReplayMismatch[Nid, Cid] extends Product with Serializable {
-  def recordedTransaction: VersionedTransaction[Nid, Cid]
-  def replayedTransaction: VersionedTransaction[Nid, Cid]
-
+final case class ReplayMismatch[Nid, Cid](
+    recordedTransaction: VersionedTransaction[Nid, Cid],
```

Comment:
this seems slightly sad, we produce somewhat worse errors now. I think this is fine since you never should see this without a malicious submitter but maybe worth considering if we should rely on structural equality at the node level but still traverse the transaction so we can pin it down to a node but let’s not do this in this transaction.